robot-name/ena/robot_name.go

```diff
@@ -0,0 +1,56 @@
+package robotname
+
+import (
+	"errors"
+	"math"
+	"math/rand"
+	"time"
+)
+
+var usedNames = make(map[string]bool)
+
+func init() {
+	rand.Seed(time.Now().UnixNano())
+}
+
+type Robot struct {
+	name string
+}
+
+func (r *Robot) Name() (string, error) {
+	const letters = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	const lettersPartLen = 2
+	const digits = "0123456789"
+	const digitsPartLen = 3
+
+	if r.name != "" {
+		return r.name, nil
+	}
+
+	allPossibleNames := int(math.Pow(float64(len(letters)), lettersPartLen) * math.Pow(float64(len(digits)), digitsPartLen))
+	if len(usedNames) == allPossibleNames {
+		return "", errors.New("All names taken")
+	}
+
+	for {
```

Comment:
This is a costly approach. For your code, the last test had to regenerate 8110729 names as they one generated it already existed. It's still nice to see that is quite fast, so well done.